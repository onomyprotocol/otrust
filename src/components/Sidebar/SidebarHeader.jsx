import React from 'react';
import styled from 'styled-components';

import { responsive } from 'theme/constants';
import { CogIcon, LogoutIcon } from './SidebarIcons';

const PrimaryIcon = styled.a`
  display: flex;
  align-items: center;
  justify-content: center;

  height: 44px;
  width: 44px;

  background-color: ${props => props.theme.colors.bgDarken};
  border: 1px solid ${props => props.theme.colors.bgHighlightBorder};
  border-radius: 8px;

  svg {
    width: 20px;
    height: 20px;
  }
  color: ${props => props.theme.colors.iconsNormal};

  cursor: pointer;

  @media screen and (max-width: ${responsive.laptop}) {
    width: 32px;
    height: 32px;

    svg {
      width: 16px;
      height: 16px;
    }
  }

  @media screen and (max-width: ${responsive.tablet}) {
    &:first-child {
      grid-column: 3/4;
    }
    &:nth-child(3) {
      grid-column: 4/5;
    }
  }

  @media screen and (max-width: ${responsive.smartphoneLarge}) {
    width: 44px;
    height: 44px;

    svg {
      width: 20px;
      height: 20px;
    }
  }

  &:hover {
    background-color: ${props => props.theme.colors.bgNormal_lighten};
  }

  &:active {
    background-color: ${props => props.theme.colors.bgNormal_darken};
  }
`;

const Header = styled.header`
  display: grid;
  grid-template-columns: 44px 72px 44px;
  grid-template-rows: auto auto;
  align-items: center;
  justify-content: center;
  gap: 32px 60px;

  height: 240px;
  padding: 40px 48px;

  background-color: ${props => props.theme.colors.bgDarken};
  border-radius: 4px;

  @media screen and (max-width: ${responsive.laptop}) {
    grid-template-columns: 32px 56px 32px;
    gap: 24px 32px;

    height: 180px;
    padding: 24px 48px;
  }

  @media screen and (max-width: ${responsive.laptopSmall}) {
    grid-template-columns: 32px 56px 32px;
    gap: 24px;

    padding: 24px;
  }

  @media screen and (max-width: ${responsive.tablet}) {
    grid-template-rows: none;
    grid-template-columns: 56px 250px 32px 32px;
    justify-items: start;
    justify-content: start;
    gap: 10px;

    height: 100px;

    background-color: ${props => props.theme.colors.bgNormal};
    border-bottom: 1px solid ${props => props.theme.colors.bgHighlightBorder};
    border-radius: 0;
  }

  @media screen and (max-width: ${responsive.tabletSmall}) {
    grid-template-columns: 56px 150px 32px 32px;

    border-radius: 0;
  }

  @media screen and (max-width: ${responsive.smartphoneLarge}) {
    grid-template-columns: 56px 1fr 44px 44px;
    gap: 16px;

    padding: 20px;
  }

  @media screen and (max-width: ${responsive.smartphoneSmall}) {
    grid-column: 1fr 44px 44px;
  }
`;

const Avatar = styled.img`
  height: 72px;
  width: 72px;

  border-radius: 12px;
  border: 3px solid ${props => props.theme.colors.bgHighlightBorder};

  @media screen and (max-width: ${responsive.laptop}) {
    height: 56px;
    width: 56px;
  }

  @media screen and (max-width: ${responsive.tablet}) {
    grid-column: 1/2;
    grid-row: 1/2;
  }

  @media screen and (max-width: ${responsive.smartphoneSmall}) {
    grid-column: none;

    display: none;
  }
`;

const AccountNumber = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 8px;

  grid-column: 1/4;

  @media screen and (max-width: ${responsive.tablet}) {
    align-items: flex-start;
    grid-column: 2/3;
    grid-row: 1/2;

    padding-left: 14px;
  }

  @media screen and (max-width: ${responsive.smartphoneSmall}) {
    grid-column: 1/2;

    padding: 0;
  }

  span {
    font-size: 16px;
    font-weight: 500;
    color: ${props => props.theme.colors.textSecondary};

    @media screen and (max-width: ${responsive.laptop}) {
      font-size: 14px;
    }
  }
`;

export default function SidebarHeader({ account, onLogout }) {
  return (
    <Header>
      <PrimaryIcon>
        <CogIcon />
      </PrimaryIcon>
      <Avatar src="https://picsum.photos/72" alt="" />
      <PrimaryIcon onClick={onLogout}>
        <LogoutIcon />
      </PrimaryIcon>
      <AccountNumber>
        <p>My Account</p>
        <span>
          {account === null
            ? '-'
            : account
            ? `${account.substring(0, 6)}...${account.substring(account.length - 4)}`
            : ''}
        </span>
      </AccountNumber>
    </Header>
  );
}
