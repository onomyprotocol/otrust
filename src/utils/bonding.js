const a = 100000000;

export function priceAtSupply(supply) {
  return (supply / a) ** 2;
}

// Bonding curve price at specified supply
export function supplyAtPrice(price) {
  //#NOM Sold = sqrt(ETH/NOM) * a
  return Math.sqrt(price) * a;
}

// NOM supply range to ETH
export function NOMsupplyETH(supplyTop, supplyBot) {
  // Integrate over curve to get amount of ETH needed to buy amount of NOM
  // ETH = a/3((supplyNOM_Top/a)^3 - (supplyNOM_Bot/a)^3)
  return (a / 3) * ((supplyTop / a) ** 3 - (supplyBot / a) ** 3);
}

// Returns Buy Quote for the purchase of NOM based on amount of ETH
// Parameters:
// Input
// amount: amount of ETH
// Output
// output: amount of NOM
export function ETHtoNOM(amount, supply) {
  const price = priceAtSupply(supply);
  // 1. Determine supply bottom
  // Price bottom is 1% above priceBondCurve
  const priceBot = price * 1.01;
  const supplyBot = supplyAtPrice(priceBot);
  // 2. Integrate over curve, and solve for supply top
  // (3*ETH/a + (supplyNOM_Bot/a)^3)^(1/3)
  const supplyTop = a * ((3 * amount) / a + supplyBot ** 3 / a ** 3) ** (1 / 3);
  // 3. Subtract supply bottom from top to get amount NOM for amount ETH
  const diff = supplyTop - supplyBot;
  return { supplyBot, supplyTop, diff };
}

// Returns Sell Quote: NOM for ETH (Dec 18)
// Parameters:
// Input
// uint256 amountNOM: amount of NOM to be sold (18 decimal)
// Output
// uint256: amount of ETH paid in Wei or ETH (18 decimal)
export function NOMtoETH(amount, supplyTop) {
  const supplyBot = supplyTop - amount;
  const diff = 0.99 * NOMsupplyETH(supplyTop, supplyBot);
  return { supplyBot, supplyTop, diff };
}
