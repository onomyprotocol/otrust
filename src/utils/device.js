const size = {
  mobileS: '320px',
  mobileM: '375px',
  mobileL: '425px',
  tablet: '768px',
  laptop: '1024px',
  desktopS: '1366px',
  desktopM: '1920px',
  desktopL: '2560px',
};

export const device = {
  mobileS: `(min-width: ${size.mobileS})`,
  mobileM: `(min-width: ${size.mobileM})`,
  mobileL: `(min-width: ${size.mobileL})`,
  tablet: `(min-width: ${size.tablet})`,
  laptop: `(min-width: ${size.laptop})`,
  desktopS: `(min-width: ${size.desktopS})`,
  desktopM: `(min-width: ${size.desktopM})`,
  desktopL: `(min-width: ${size.desktopL})`,
};
