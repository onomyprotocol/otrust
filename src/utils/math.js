import { BigNumber } from 'bignumber.js';

export function chopFloat(value, precision) {
  return Math.round(value * 10 ** precision) / 10 ** precision;
}

export function randomInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1) + min);
}

export function truncate(str, maxDecimalDigits) {
  if (str.includes('.')) {
    const parts = str.split('.');
    return parts[0] + '.' + parts[1].slice(0, maxDecimalDigits);
  }
  return str;
}

export function format18(bignumber) {
  return bignumber.div(new BigNumber(10 ** 18));
}

export function parse18(bignumber) {
  return bignumber.times(new BigNumber(10 ** 18));
}

export var isNumber = function isNumber(value) {
  return typeof value === 'number' && isFinite(value);
};
