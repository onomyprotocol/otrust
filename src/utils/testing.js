import { ThemeProvider } from 'styled-components';
import ApolloClient, { InMemoryCache } from 'apollo-boost';
import { ApolloProvider } from '@apollo/client';
import { render } from '@testing-library/react';
import { BigNumber } from 'bignumber.js';

import { darkNew } from 'Theme/theme';
import { ChainContext } from '../context/chain/ChainContext';
import { ExchangeContext } from '../context/exchange/ExchangeContext';

const client = new ApolloClient({
  uri: process.env.REACT_APP_GRAPHQL_ENDPOINT,
  cache: new InMemoryCache(),
});

export const renderWithTheme = (Component, props, children) => {
  if (children) {
    return render(
      <ThemeProvider theme={darkNew}>
        <Component {...props}>{children}</Component>
      </ThemeProvider>,
    );
  }
  return render(
    <ThemeProvider theme={darkNew}>
      <Component {...props} />
    </ThemeProvider>,
  );
};

export const renderWithContext = (Component, props) => {
  return render(
    <ThemeProvider theme={darkNew}>
      <ApolloProvider client={client}>
        <ChainContext.Provider value={{ supplyNOM: new BigNumber(0), theme: darkNew }}>
          <ExchangeContext.Provider value={{ askAmount: 0, bidAmount: 0, bidDenom: 'strong' }}>
            <Component {...props} />
          </ExchangeContext.Provider>
        </ChainContext.Provider>
      </ApolloProvider>
    </ThemeProvider>,
  );
};
