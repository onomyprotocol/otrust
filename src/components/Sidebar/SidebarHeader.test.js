import { cleanup } from '@testing-library/react';

import SidebarHeader from './SidebarHeader';
import { renderWithTheme } from '../../utils/testing';

const testProps = {
  setDenom: jest.fn(),
  account: '123456789012',
};

describe('Given the SidebarHeader component', () => {
  describe('when the component is rendered and account is defined', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(SidebarHeader, testProps);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when the component is rendered and account is null', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(SidebarHeader, { account: null });
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
