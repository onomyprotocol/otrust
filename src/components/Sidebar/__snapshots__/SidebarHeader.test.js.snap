// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the SidebarHeader component when the component is rendered and account is defined should match the snapshot 1`] = `
<DocumentFragment>
  <header
    class="sc-gtsrHT cWvxos"
  >
    <a
      class="sc-bdnxRM fDEryd"
    >
      <svg
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h20v20H0z"
          fill="none"
        />
        <path
          d="M10.519 17.348H9.484a1.538 1.538 0 01-1.582-1.526 1.417 1.417 0 00-1.4-1.355 1.176 1.176 0 00-.669.172 1.61 1.61 0 01-.804.221 1.58 1.58 0 01-1.354-.774l-.525-.894a1.492 1.492 0 01-.021-1.524 1.569 1.569 0 01.608-.588 1.23 1.23 0 00.48-.476 1.327 1.327 0 00.141-1.016 1.367 1.367 0 00-.629-.845 1.522 1.522 0 01-.567-2.107l.509-.876a1.58 1.58 0 012.145-.565 1.41 1.41 0 00.668.167 1.458 1.458 0 001.245-.684 1.145 1.145 0 00.176-.655 1.331 1.331 0 01.21-.789 1.64 1.64 0 011.346-.774h1.068a1.6 1.6 0 011.359.774 1.326 1.326 0 01.205.789 1.15 1.15 0 00.175.655 1.462 1.462 0 001.245.68 1.433 1.433 0 00.673-.167 1.571 1.571 0 012.138.565l.509.878a1.509 1.509 0 01-.563 2.107 1.362 1.362 0 00-.635.847 1.318 1.318 0 00.148 1.014 1.169 1.169 0 00.472.476 1.717 1.717 0 01.616.588 1.512 1.512 0 01-.015 1.526l-.533.894a1.562 1.562 0 01-2.153.551 1.207 1.207 0 00-.669-.172 1.414 1.414 0 00-1.408 1.355 1.534 1.534 0 01-1.574 1.528zm-.51-9.543a2.136 2.136 0 00-2.161 2.107A2.057 2.057 0 008.48 11.4a2.181 2.181 0 001.528.611 2.1 2.1 0 001.521-3.591 2.142 2.142 0 00-1.52-.615z"
          fill="#8d89a3"
        />
      </svg>
    </a>
    <img
      alt=""
      class="sc-dlnjwi gppAIF"
      src="https://picsum.photos/72"
    />
    <a
      class="sc-bdnxRM fDEryd"
    >
      <svg
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h20v20H0z"
          fill="none"
        />
        <path
          d="M18.643 10.674l-3.367 3.3a.987.987 0 01-1.4-1.392l1.137-1.129c.259-.257.173-.465-.192-.465H10.3a.984.984 0 110-1.968h4.525c.363 0 .451-.208.192-.466L13.88 7.425a.987.987 0 011.4-1.392l3.359 3.262a.957.957 0 01.004 1.379zM5.365 14.936h1.977a.987.987 0 110 1.974H4.379a1.317 1.317 0 01-1.314-1.316V4.41a1.317 1.317 0 011.314-1.316h2.963a.987.987 0 110 1.974H5.365a.327.327 0 00-.33.328v9.214a.329.329 0 00.33.326z"
          fill="#8d89a3"
          fill-rule="evenodd"
        />
      </svg>
    </a>
    <div
      class="sc-hKFxyN dqEzzQ"
    >
      <p>
        My Account
      </p>
      <span>
        123456...9012
      </span>
    </div>
  </header>
</DocumentFragment>
`;

exports[`Given the SidebarHeader component when the component is rendered and account is null should match the snapshot 1`] = `
<DocumentFragment>
  <header
    class="sc-gtsrHT cWvxos"
  >
    <a
      class="sc-bdnxRM fDEryd"
    >
      <svg
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h20v20H0z"
          fill="none"
        />
        <path
          d="M10.519 17.348H9.484a1.538 1.538 0 01-1.582-1.526 1.417 1.417 0 00-1.4-1.355 1.176 1.176 0 00-.669.172 1.61 1.61 0 01-.804.221 1.58 1.58 0 01-1.354-.774l-.525-.894a1.492 1.492 0 01-.021-1.524 1.569 1.569 0 01.608-.588 1.23 1.23 0 00.48-.476 1.327 1.327 0 00.141-1.016 1.367 1.367 0 00-.629-.845 1.522 1.522 0 01-.567-2.107l.509-.876a1.58 1.58 0 012.145-.565 1.41 1.41 0 00.668.167 1.458 1.458 0 001.245-.684 1.145 1.145 0 00.176-.655 1.331 1.331 0 01.21-.789 1.64 1.64 0 011.346-.774h1.068a1.6 1.6 0 011.359.774 1.326 1.326 0 01.205.789 1.15 1.15 0 00.175.655 1.462 1.462 0 001.245.68 1.433 1.433 0 00.673-.167 1.571 1.571 0 012.138.565l.509.878a1.509 1.509 0 01-.563 2.107 1.362 1.362 0 00-.635.847 1.318 1.318 0 00.148 1.014 1.169 1.169 0 00.472.476 1.717 1.717 0 01.616.588 1.512 1.512 0 01-.015 1.526l-.533.894a1.562 1.562 0 01-2.153.551 1.207 1.207 0 00-.669-.172 1.414 1.414 0 00-1.408 1.355 1.534 1.534 0 01-1.574 1.528zm-.51-9.543a2.136 2.136 0 00-2.161 2.107A2.057 2.057 0 008.48 11.4a2.181 2.181 0 001.528.611 2.1 2.1 0 001.521-3.591 2.142 2.142 0 00-1.52-.615z"
          fill="#8d89a3"
        />
      </svg>
    </a>
    <img
      alt=""
      class="sc-dlnjwi gppAIF"
      src="https://picsum.photos/72"
    />
    <a
      class="sc-bdnxRM fDEryd"
    >
      <svg
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h20v20H0z"
          fill="none"
        />
        <path
          d="M18.643 10.674l-3.367 3.3a.987.987 0 01-1.4-1.392l1.137-1.129c.259-.257.173-.465-.192-.465H10.3a.984.984 0 110-1.968h4.525c.363 0 .451-.208.192-.466L13.88 7.425a.987.987 0 011.4-1.392l3.359 3.262a.957.957 0 01.004 1.379zM5.365 14.936h1.977a.987.987 0 110 1.974H4.379a1.317 1.317 0 01-1.314-1.316V4.41a1.317 1.317 0 011.314-1.316h2.963a.987.987 0 110 1.974H5.365a.327.327 0 00-.33.328v9.214a.329.329 0 00.33.326z"
          fill="#8d89a3"
          fill-rule="evenodd"
        />
      </svg>
    </a>
    <div
      class="sc-hKFxyN dqEzzQ"
    >
      <p>
        My Account
      </p>
      <span>
        -
      </span>
    </div>
  </header>
</DocumentFragment>
`;
